mantle/cmd/kola/devshell.go

```diff
@@ -0,0 +1,303 @@
+ExecStart=/bin/echo ready
+StandardOutput=file:/dev/virtio-ports/%s
+[Install]
+RequiredBy=multi-user.target`, readinessSignalChan, readinessSignalChan)
+	signalPoweroffUnit := fmt.Sprintf(`[Unit]
+Requires=dev-virtio\\x2dports-%s.device
+OnFailure=emergency.target
+OnFailureJobMode=isolate
+Conflicts=reboot.target
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStop=/bin/echo poweroff
+StandardOutput=file:/dev/virtio-ports/%s
+[Install]
+WantedBy=multi-user.target`, readinessSignalChan, readinessSignalChan)
+
+	devshellConfig := v3types.Config{
+		Ignition: v3types.Ignition{
+			Version: "3.0.0",
+		},
+		Passwd: v3types.Passwd{
+			Users: []v3types.PasswdUser{
+				{
+					Name: "core",
+					SSHAuthorizedKeys: []v3types.SSHAuthorizedKey{
+						sshPubKey,
+					},
+				},
+			},
+		},
+		Systemd: v3types.Systemd{
+			Units: []v3types.Unit{
+				{
+					Name:     "coreos-devshell-signal-ready.service",
+					Contents: &signalReadyUnit,
+					Enabled:  util.BoolToPtr(true),
+				},
+				{
+					Name:     "coreos-devshell-signal-poweroff.service",
+					Contents: &signalPoweroffUnit,
+					Enabled:  util.BoolToPtr(true),
+				},
+			},
+		},
+	}
+	confm := v3.Merge(*conf, devshellConfig)
+	conf = &confm
+
+	readyChan, err := builder.VirtioChannelRead(readinessSignalChan)
+	if err != nil {
+		return err
+	}
+	readyReader := bufio.NewReader(readyChan)
+
+	if err := builder.SetConfig(*conf, kola.Options.IgnitionVersion == "v2"); err != nil {
+		return errors.Wrapf(err, "rendering config")
+	}
+
+	builder.InheritConsole = devshellConsole
+	inst, err := builder.Exec()
+	if err != nil {
+		return err
+	}
+	defer inst.Destroy()
+
+	if devshellConsole {
+		return inst.Wait()
+	}
+
+	qemuWaitChan := make(chan error)
+	errchan := make(chan error)
+	readychan := make(chan struct{})
+	go func() {
+		buf, err := inst.WaitIgnitionError()
+		if err != nil {
+			errchan <- err
+		} else {
+			// TODO parse buf and try to nicely render something
+			if buf != "" {
+				errchan <- platform.ErrInitramfsEmergency
+			}
+		}
+	}()
+	go func() {
+		qemuWaitChan <- inst.Wait()
+	}()
+	go func() {
+		readyMsg, err := readTrimmedLine(readyReader)
+		if err != nil {
+			errchan <- err
+		}
+		if readyMsg != "ready" {
+			errchan <- fmt.Errorf("Unexpected ready message: %s", readyMsg)
+		}
+		var s struct{}
+		readychan <- s
+	}()
+
+	select {
+	case err := <-errchan:
+		if err == platform.ErrInitramfsEmergency {
+			return fmt.Errorf("instance failed in initramfs; try rerunning with --devshell-console")
+		}
+		return err
+	case err := <-qemuWaitChan:
+		return errors.Wrapf(err, "qemu exited before setup")
+	case _ = <-readychan:
+		fmt.Println("virtio: connected")
+	}
+
+	var ip string
+	err = util.Retry(6, 5*time.Second, func() error {
+		var err error
+		ip, err = inst.SSHAddress()
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "awaiting ssh address")
+	}
+
+	sshConfigPath := filepath.Join(tmpd, "ssh-config")
```

Comment:
Could probably dedup a lot of the below with `appendSSH` a function that just spits out a full SSH config given an IP. But not a blocker.

Reply:
Will consider as followup though what I really want to do is use a virtio-channel instead of TCP and then we can skip some overhead, and that'd make these cases different.